README.md
# [supergiovane](http://supergiovane.tk/#/supergiovane)

[![NPM version](https://img.shields.io/npm/v/supergiovane.svg)](https://www.npmjs.com/package/supergiovane)
[![Linux Status](https://img.shields.io/travis/hex7c0/supergiovane.svg?label=linux)](https://travis-ci.org/hex7c0/supergiovane)
[![Dependency Status](https://img.shields.io/david/hex7c0/supergiovane.svg)](https://david-dm.org/hex7c0/supergiovane)
[![Coveralls](https://img.shields.io/coveralls/hex7c0/supergiovane.svg)](https://coveralls.io/r/hex7c0/supergiovane)

[![supergiovane logo](https://raw.githubusercontent.com/hex7c0/supergiovane/master/public/img/sp.jpg)](http://supergiovane.tk)

Website for get info about npm module.
Build badge with all versions found (using https://github.com/badges/shields) [![versions](http://supergiovane.tk/supergiovane/badge.svg)](http://supergiovane.tk/#/supergiovane)
Write with [Bootstrap](http://getbootstrap.com/) & [Angular](https://angularjs.org/). And of course [node](http://nodejs.org/)

## Installation

Install through NPM

```bash
npm install supergiovane
```
or
```bash
git clone git://github.com/hex7c0/supergiovane.git
```

## API

```js
var supergiovane = require('supergiovane');

supergiovane(options);
```

### supergiovane(options)

#### options

 - `env` - **String** Selected environment (development, production, test) *(default "production")*
 - `host` - **String** Which host accept connection from web *(default "127.0.0.1")*
 - `port` - **Integer** Which port accept connection from web *(default "3000")*
 - `referer` - **String** Which referer send data *(default "http://127.0.0.1")*
 - `dir` - **String** Absolute path of web files directory *(default "/public")*
 - `logger` - **Object | false** Configuration for [`logger-request`](https://github.com/hex7c0/logger-request) or disable *(default "standard")*
 - `timeout` - **Object | false** Configuration for [`timeout-request`](https://github.com/hex7c0/timeout-request) or disable *(default "standard")*
 - `sitemap` - **Object | false** Configuration for [`express-sitemap`](https://github.com/hex7c0/express-sitemap) or disable *(default "standard")*
 - `vhost` - **Object | false** Configuration for [`top-vhost`](https://github.com/hex7c0/top-vhost) or disable *(default "disabled")*
 - `signature` - **Object | false** Configuration for [`server-signature`](https://github.com/hex7c0/server-signature) or disable *(default "disabled")*
 - `cache` - **Number | false** Flag for store last "Number" results in cache *(default "6")*
 - `flush` - **Number** Flush cache (if enabled) after milliseconds *(default "86400000")*
 - `fork` - **Number** Number of forks (see cluster module), for single thread use "development" `env` *(default "cpus")*
 - `max` - **Number | String** Restart max number of crashed forks (use String for unlimited restart) *(default "no restart")*
 - `debug` - **String | false** Write debug information to file or disable *(default "debug.log")*
 - `task` - **Number | String | false** Related to [`task-manager`](https://github.com/hex7c0/task-manager) *(default "false")*
 - `mamma` - **Number | String | false** Related to [`mamma`](https://github.com/hex7c0/mamma) `createClient` *(default "false")*

## Examples

Take a look at my [examples](https://github.com/hex7c0/supergiovane/tree/master/examples)

### [License GPLv3](http://opensource.org/licenses/GPL-3.0)


History.md
v1.7.2 / 2015-07-06
==================

  * Improve Iojs test
  * _Deps_:
    * Update `express`@4.13.1

v1.7.1 / 2015-07-05
==================

  * _Deps_:
    * Update `compression-zlib`@2.1.0
    * Update `express`@4.13.0
    * Update `express-sitemap`@1.5.6
    * Update `logger-request`@3.3.5
    * Update `semver`@4.3.6
  * _Web_:
    * Update `bootstrap`@3.3.5
    * Update `angular`@1.4.1

v1.7.0 / 2015-05-02
==================

  * _Deps_:
    * Update `compression-zlib`@2.0.2
    * Update `express`@4.12.3
    * Update `express-sitemap`@1.5.5
    * Update `mamma`@0.2.0
    * Update `logger-request`@3.3.2
    * Update `semver`@4.3.3
    * Update `server-signature`@1.2.0
    * Update `timeout-request`@1.4.0
    * Update `task-manager`@1.3.0
  * _Web_:
    * Update `bootstrap`@3.3.4
    * Update `angular`@1.4.0-rc.1
    * Update `jquery`@2.1.4

v1.6.15 / 2015-03-01
==================

  * _Deps_:
    * Add `setheaders`@0.1.1
    * Add `compression-zlib`@2.0.1
    * Update `express`@4.12.0
    * Update `express-sitemap`@1.5.3
    * Update `logger-request`@3.3.1
    * Update `server-signature`@1.1.0
    * Update `semver`@4.3.1
  * _Web_:
    * Update `bootstrap`@3.3.2
    * Update `angular`@1.4.0-beta.5

v1.6.14 / 2015-01-18
==================

  * _Deps_:
    * Add `mamma`@0.1.1
    * Update `express`@4.11.0
    * Update `server-signature`@1.0.9
    * Update `task-manager`@1.2.0
    * Update `express-sitemap`@1.4.0
  * _Web_:
    * Update `angular`@1.3.9

v1.6.13 / 2015-01-11
==================

  * Fix `console.log` bug when shell is closed
  * Update `server-signature`@1.0.8

v1.6.12 / 2015-01-10
==================

  * Add uncaughtException logic

v1.6.11 / 2015-01-05
==================

  * Cluster gracefully exit
  * _Deps_:
    * Update `logger-request`@3.2.7
    * Update `task-manager`@1.1.0
    * Update `compression`@1.3.0
    * Update `express`@4.10.7

v1.6.10 / 2014-12-2
==================

  * Fix options parsing

v1.6.9 / 2014-12-24
==================

  * _Deps_:
    * Update `logger-request`@3.2.5
    * Update `semver`@4.2.0
  * _Web_:
    * Update `angular`@1.3.8
    * Update `jquery`@2.1.3

v1.6.8 / 2014-12-15
==================

  * Update `express`@4.10.6

v1.6.7 / 2014-12-11
==================

  * Fix breaking change on `http.get`

v1.6.6 / 2014-12-11
==================

  * Fix http timeout
  * _Deps_:
    * Update `compression`@1.2.2
    * Update `express`@4.10.5
  * _Web_:
    * Break commit add `angular-sanitize`

v1.6.5 / 2014-12-10
==================

  * Update web interface with new npm website
  * _Deps_:
    * Remove `lusca`
  * _Web_:
    * Remove `angular-sanitize`
    * Update `angular`@1.3.6

v1.6.4 / 2014-12-03
==================

  * _Deps_:
    * Update `logger-request`@3.2.3
    * Update `timeout-request`@1.3.0
  * _Web_:
    * Update `angular`@1.3.5

v1.6.3 / 2014-11-28
==================

  * _Deps_:
    * Update `compression`@1.2.1
    * Update `express`@4.10.4
  * _Web_:
    * Update `angular`@1.3.4

v1.6.2 / 2014-11-22
==================

  * _Deps_:
    * Update `express-sitemap`@1.3.8
    * Update `logger-request`@3.2.2
    * Update `server-signature`@1.0.7
  * _Web_:
    * Update `bootstrap`@3.3.1
    * Update `angular`@1.3.3

v1.6.1 / 2014-11-11
==================

  * Web Fix
  * Update `logger-request`@3.2.0
  * Update `timeout-request`@1.2.1
  * Update `express`@4.10.2
  * Update `angular`@1.3.2

v1.6.0 / 2014-11-04
==================

  * _Deps_:
    * Update `logger-request`@3.1.1
    * Update `server-signature`@1.0.5
    * Update `timeout-request`@1.2.0
    * Update `express`@4.10.1
  * _Web_:
    * Update `bootstrap`@3.3.0
    * Update `angular`@1.3.1

v1.5.24 / 2014-10-28
==================

  * `jshint`
  * Update `logger-request`@3.0.13
  * Update `express-sitemap`@1.3.7
  * Update `task-manager`@1.0.3
  * Update `server-signature`@1.0.4
  * Update `timeout-request`@1.1.7

v1.5.23 / 2014-10-25
==================

  * Update `logger-request`@3.0.11
  * Fix `express`@4.10.0

v1.5.22 / 2014-10-20
==================

  * Fix `express`@4.9.8

v1.5.21 / 2014-10-20
==================

  * Fix `lusca`@1.0.2

v1.5.20 / 2014-10-20
==================

  * Update `timeout-request`@1.0.3
  * Update `compression`@1.2.0
  * Update `express`@4.9.8
  * Update `lusca`@1.0.2
  * Update `semver`@4.1.0

v1.5.19 / 2014-10-14
==================

  * Update `task-manager`@1.0.2
  * Update `compression`@1.1.1
  * Update `angular`@1.3.0

v1.5.18 / 2014-10-12
==================

  * Update `express`@4.9.7

v1.5.17 / 2014-10-10
==================

  * Update `logger-request`@3.0.10
  * Update `express`@4.9.6
  * Update `angular`@1.3.0-rc.5

v1.5.16 / 2014-10-04
==================

  * Remove jsdoc
  * `package.json` min

v1.5.15 / 2014-10-02
==================

  * Remove `top-vhost`@1.5.2
  * Update `logger-request`@3.0.8
  * Update `task-manager`@1.0.0
  * Update `semver`@4.0.3
  * Fix parsing options 
  * Update `angular`@1.3.0-rc.4

v1.5.14 / 2014-09-25
==================

  * Update `express`@4.9.5

v1.5.13 / 2014-09-25
==================

  * Update `task-manager`@0.1.0
  * Update `semver`@4.0.0
  * Update `angular`@1.3.0-rc.3

v1.5.12 / 2014-09-20
==================

  * Update `express`@4.9.4

v1.5.11 / 2014-09-19
==================

  * Update `express`@4.9.3

v1.5.10 / 2014-09-18
==================

  * Update `logger-request`@3.0.7
  * Update `express`@4.9.2

v1.5.9 / 2014-09-17
==================

  * Update `express`@4.9.1
  * Update `angular`@1.3.0-rc.2

v1.5.8 / 2014-09-16
==================

  * Add "task" (options)
  * Using `task-manager`@0.0.2

v1.5.7 / 2014-09-15
==================

  * Update `top-vhost`@1.7.5
  * Better error handling
  * Return json only
  * Remove `csurf`@1.6.1
  * Remove `cookie-parser`@1.3.3

v1.5.6 / 2014-09-11
==================

  * Update `angular`@1.3.0-rc.1

v1.5.5 / 2014-08-09
==================

  * Update `express`@4.9.0

v1.5.4 / 2014-08-08
==================

  * Remove useless controls

test/http.js
'use strict';
/**
 * @file http test
 * @module supergiovane
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @license GPLv3
 */

/*
 * initialize module
 */
var supergiovane = require('..');
var request = require('supertest');
var assert = require('assert');
var fs = require('fs');

/*
 * test module
 */
describe('http', function() {

  var app;
  var r = __dirname + '/r.txt';
  var s = __dirname + '/s.xml';

  before(function(done) {

    app = supergiovane({
      env: 'test',
      sitemap: {
        robots: r,
        sitemap: s
      },
      debug: 'debug2.log'
    });
    done();
  });

  describe('200 status code', function() {

    var cache;

    it('should get index', function(done) {

      request(app).get('/').expect(200, done);
    });
    it('should get package', function(done) {

      request(app).get('/supergiovane/').set('Referer', 'http://127.0.0.1')
      .expect(200).end(function(err, res) {

        assert.equal(err, null);
        assert.deepEqual(res.statusCode, 200);
        cache = JSON.parse(res.text);
        assert.deepEqual(cache.name, 'supergiovane');
        assert.deepEqual(cache.versions['0.0.1'].main, 'index.min.js');
        assert.deepEqual(cache.license, 'GPL-3.0');
        done();
      });
    });
    it('should get cached', function(done) {

      request(app).get('/supergiovane/').set('Referer', 'http://127.0.0.1')
      .expect(202).end(function(err, res) {

        if (err) throw err;
        var j = JSON.parse(res.text);
        assert.deepEqual(cache, j);
        done();
      });
    });
    it('shouldn\'t get package without refer', function(done) {

      request(app).get('/supergiovane/').expect(301, done);
    });
    it('shouldn\'t get package different refer', function(done) {

      request(app).get('/supergiovane/').set('Referer', 'mah')
      .expect(301, done);
    });
  });

  describe('404 status code', function() {

    it('shouldn\'t get static', function(done) {

      request(app).get('/index.html').set('Referer', 'http://127.0.0.1')
      .expect(404, done);
    });
    it('shouldn\'t get wrong package', function(done) {

      request(app).get('/supergiovane_qwertyuiop/').set('Referer',
        'http://127.0.0.1').expect(404, done);
    });
  });

  describe('files', function() {

    it('should remove logger', function(done) {

      fs.unlink('route.log', done);
    });
    it('should remove robots', function(done) {

      fs.unlink(r, done);
    });
    it('should remove sitemap', function(done) {

      fs.unlink(s, done);
    });
    it('should remove debug', function(done) {

      fs.unlink('debug2.log', done);
    });
  });
});


test/extra.js
'use strict';
/**
 * @file extra test
 * @module supergiovane
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @license GPLv3
 */

/*
 * initialize module
 */
var supergiovane = require('..');
var unlink = require('fs').unlink;
var request = require('supertest');
var assert = require('assert');

/*
 * test module
 */
describe('extra', function() {

  var app;
  var cache1;
  var cache2;

  before(function(done) {

    app = supergiovane({
      env: 'test',
      debug: false,
      sitemap: false,
      cache: 2,
      task: 'test.sock'
    });
    done();
  });

  describe('wrong version', function() {

    it('should return 404 for missing version', function(done) {

      request(app).get('/supergiovane/0.0.2')
      .set('Referer', 'http://127.0.0.1').expect(404, done);
    });
    it('should return 404 for misconfigurated version', function(done) {

      request(app).get('/supergiovane/0.ciao.0').set('Referer',
        'http://127.0.0.1').expect(404, done);
    });
  });

  describe('first attempt normal', function() {

    it('should return badge', function(done) {

      request(app).get('/supergiovane/badge.svg').expect(200).expect(
        'Content-Type', 'image/svg+xml; charset=utf-8').end(function(err, res) {

        assert.equal(err, null);
        if (/^<svg xmlns="http:\/\/www.w3.org\/2000\/svg"/.test(res.text)) {
          cache1 = res.text;
          done();
        } else {
          throw new Error('not svg');
        }
      });
    });
    it('should return specific version', function(done) {

      request(app).get('/supergiovane/1.0.0')
      .set('Referer', 'http://127.0.0.1').expect(200).expect('Content-Type',
        'application/json; charset=utf-8').end(
        function(err, res) {

          assert.equal(err, null);
          cache2 = JSON.parse(res.text);
          assert.deepEqual(cache2.version, '1.0.0', 'version');
          assert.deepEqual(cache2.dist.shasum,
            '9bdc48303b50f2c7b2b28102f45f25dc3c2307ce', 'sha');
          done();
        });
    });
  });

  describe('second attempt normal', function() {

    it('should return badge', function(done) {

      request(app).get('/supergiovane/badge.svg').expect(202).expect(
        'Content-Type', 'image/svg+xml; charset=utf-8').end(function(err, res) {

        assert.equal(err, null);
        assert.equal(err, null);
        assert.deepEqual(cache1, cache1);
        assert.notDeepEqual(cache2, cache1);
        done();
      });
    });
    it('should return specific version', function(done) {

      request(app).get('/supergiovane/1.0.0')
      .set('Referer', 'http://127.0.0.1').expect(202).expect('Content-Type',
        'application/json; charset=utf-8').end(function(err, res) {

        assert.equal(err, null);
        assert.deepEqual(cache2, cache2);
        assert.notDeepEqual(cache1, cache2);
        done();
      });
    });
  });

  describe('files', function() {

    it('should remove domain socket', function(done) {

      unlink('test.sock', done);
    });
  });
});


test/options.js
'use strict';
/**
 * @file options test
 * @module supergiovane
 * @subpackage test
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @license GPLv3
 */

/*
 * initialize module
 */
var supergiovane = require('..');
var request = require('supertest');
var assert = require('assert');
var fs = require('fs');

/*
 * test module
 */
describe('options', function() {

  var app;
  var r = __dirname + '/r.txt';
  var s = __dirname + '/s.xml';

  before(function(done) {

    app = supergiovane({
      port: 3001,
      env: 'test',
      dir: __dirname + '/',
      referer: 'boh',
      logger: false,
      sitemap: false,
      signature: {
        token: 'Full'
      },
      debug: false
    });
    done();
  });

  describe('200 status code', function() {

    it('should get index', function(done) {

      request(app).get('/').expect(200, done);
    });
    it('should get package', function(done) {

      request(app).get('/supergiovane/').expect('Server',
        /Nodejs\/[0-9]{1,2}.[0-9]{1,2}.[0-9]{1,2} \(/).set('Referer', 'boh')
      .expect(200).end(function(err, res) {

        assert.equal(err, null);
        assert.deepEqual(res.statusCode, 200);
        var j = JSON.parse(res.text);
        assert.deepEqual(j.name, 'supergiovane');
        assert.deepEqual(j.versions['0.0.1'].main, 'index.min.js');
        assert.deepEqual(j.license, 'GPL-3.0');
        done();
      });
    });
  });

  describe('301 status code', function() {

    it('shouldn\'t get package, because different refer', function(done) {

      request(app).get('/supergiovane/').set('Referer', 'mah')
      .expect(301, done);
    });
  });

  describe('files', function() {

    it('shouldn\'t exist logger', function(done) {

      assert.equal(fs.existsSync('route.log'), false);
      done();
    });
    it('shouldn\'t exist robots', function(done) {

      assert.equal(fs.existsSync(r), false);
      done();
    });
    it('shouldn\'t exist sitemap', function(done) {

      assert.equal(fs.existsSync(s), false);
      done();
    });
    it('shouldn\'t exist debug', function(done) {

      assert.equal(fs.existsSync('debug.log'), false);
      done();
    });
  });
});


Gruntfile.js
'use strict';
/**
 * @file gruntfile
 * @subpackage main
 * @version 0.0.1
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

module.exports = function(grunt) {

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    banner: '/*\n' + ' * <%= pkg.name %> v<%= pkg.version %>\n'
      + ' * (c) <%= pkg.author.name %> <%= pkg.homepage %>\n'
      + ' * Licensed under <%= pkg.license %>\n' + ' */\n',

    clean: [ 'index.min.js', 'min/**/*', 'public/**/*.min.*' ],

    uglify: {
      target: {
        options: {
          mangle: false,
          beautify: true
        },
        files: [ {
          expand: true,
          src: 'lib/**/*.js',
          dest: 'min'
        }, {
          expand: true,
          src: 'module/**/*.js',
          dest: 'min'
        }, {
          'index.min.js': 'index.js'
        } ]
      },
      script: {
        options: {
          mangle: false,
          banner: '<%= banner %>'
        },
        files: {
          'public/js/script.min.js': 'public/js/script.js'
        }
      }
    },

    cssmin: {
      css: {
        options: {
          banner: '<%= banner %>'
        },
        files: {
          'public/css/style.min.css': 'public/css/style.css'
        }
      }
    },

    htmlmin: {
      html: {
        options: {
          removeComments: true,
          collapseWhitespace: true,
          minifyJS: true,
          minifyCSS: true
        },
        files: {
          'public/index.min.html': 'public/index.html'
        }
      }
    },

    jshint: {
      options: {
        curly: true,
        indent: 2,
        quotmark: 'single',
        undef: true,
        unused: true,
        strict: true,
        node: true,
        // relax
        laxbreak: true,
        loopfunc: true,
        shadow: true
      },
      target: {
        src: [ 'lib/**/*.js', 'module/**/*.js', 'index.js' ]
      },
      web: {
        options: {
          // override
          unused: false,
          strict: false,
          node: false,
          // web
          predef: [ 'angular', 'console' ],
          browser: true,
          jquery: true
        },
        files: {
          src: 'public/js/script.js'
        }
      }
    },

    shell: {
      options: {
        failOnError: false
      },
      docs: {
        command: 'jsdoc ./lib/*.js ./module/*.js -c .jsdoc.json'
      }
    },

    endline: {
      target: {
        options: {
          except: [ 'node_modules', 'bower_components' ],
          replaced: true
        },
        files: [ {
          src: './**/*.js'
        }, {
          src: './**/*.css'
        }, {
          src: './**/*.html'
        } ]
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-cssmin');
  grunt.loadNpmTasks('grunt-contrib-htmlmin');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-endline');

  grunt.registerTask('lint', [ 'jshint' ]);
  grunt.registerTask('html', [ 'cssmin', 'htmlmin' ]);
  grunt.registerTask('min', [ 'clean', 'html', 'uglify', 'endline' ]);
  grunt.registerTask('default', [ 'lint', 'min' ]);

  return;
};


index.js
'use strict';
/**
 * @file supergiovane main
 * @module supergiovane
 * @subpackage main
 * @version 1.7.0
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var cluster = require('cluster');
var http = require('http');
var resolve = require('path').resolve;
var status = http.STATUS_CODES;
var express = require('express');
var logger = require('logger-request');
var semver = require('semver');
var setHeader = require('setheaders').setWritableHeader;
// load
var VERSION = JSON.parse(require('fs')
.readFileSync(__dirname + '/package.json'));
VERSION = VERSION.name + '@' + VERSION.version;
var debug = function() {

  return;
};

/*
 * functions
 */
/**
 * starting point
 * 
 * @function bootstrap
 * @param {Object} my - user cfg
 * @return {Object}
 */
function bootstrap(my) {

  // setting
  var app = express();
  app.set('env', my.env);
  if (my.env == 'production') {
    app.enable('view cache');
  } else {
    app.disable('view cache');
  }
  app.enable('case sensitive routing');
  app.enable('trust proxy');
  app.disable('x-powered-by');

  // middleware
  if (my.logger) {
    app.use(logger(my.logger));
  }
  if (my.timeout) {
    app.use(require('timeout-request')(my.timeout));
  }
  if (my.signature) {
    app.use(require('server-signature')(my.signature));
  }
  if (my.sitemap) {
    require('express-sitemap')(my.sitemap).toFile();
  }
  if (my.mamma) {
    require('mamma').createClient(my.mamma, process.pid + ':supergiovane').on(
      'error', function(err) {

        return debug('cluster', {
          pid: process.pid,
          status: 'mamma',
          error: err.message
        });
      });
  }
  app.use(require('compression-zlib')());

  var cache;
  if (my.cache) {
    /**
     * save cache
     * 
     * @function caches
     * @param {String} bod - output body
     * @param {String} hash - hash
     * @param {String} cont - set content
     * @return {Object}
     */
    cache = function(bod, hash, cont) {

      if (STORY[hash] === undefined) {
        var c = 0, old = Date.now(), last;
        for ( var pr in STORY) {
          c++;
          if (STORY[pr].time < old) {
            last = pr;
            old = STORY[last].time;
          }
        }
        if (c >= my.cache) {
          delete STORY[last];
        }
        STORY[hash] = {
          content: cont,
          body: bod,
          time: Date.now()
        };
      }
      return;
    };
  } else {
    cache = function() {

      return;
    };
  }

  // cfg
  // http.globalAgent.maxSockets = Math.pow(my.fork, 2); // by default set to Infinity on iojs 1.0
  var httpAgent = new http.Agent({
    keepAlive: true
  });
  var STORY = Object.create(null);
  var index = resolve(my.dir + 'index.min.html');

  // routing
  /**
   * http request (no client ajax due browser security limitation)
   * 
   * @function
   * @param {Object} req - client request
   * @param {Object} res - response to client
   * @param {next} next - next callback
   */
  app.get('/:pkg/:extra?', function(req, res, next) {

    var version = '/';
    var r = req.headers.referer || req.headers.referrer;
    var p = req.params.pkg;
    var e = req.params.extra || '';
    var s = req.query.style ? '?style=' + req.query.style : '';
    var hash = p + e + s;

    // checkpoint
    if (my.referer.test(r) === false && e !== 'badge.svg') {
      return res.redirect(301, my.referer.source);
    } else if (my.cache !== false && STORY[hash]) {
      // flush cache after 1 day
      if (Date.now() - STORY[hash].time > my.flush) {
        delete STORY[hash];
      }
      return setHeader(res, 'Content-Type', STORY[hash].content) === true ? res
      .status(202).send(STORY[hash].body) : null;
    }

    if (e !== '') { // extra information
      if (e === 'badge.svg') {
        // pass
      } else if (semver.valid(e)) {
        version += e;
      } else {
        return next(new Error(status[404]));
      }
    }

    return http.get({
      host: 'registry.npmjs.org',
      path: '/' + p + version,
      agent: httpAgent,
      headers: {
        'User-Agent': VERSION
      }
    }, function(inp) {

      if (inp.statusCode !== 200) {
        return next(new Error(status[404]));
      }
      var body = new Buffer(0);

      inp.on('data', function(chunk) {

        body += chunk; // buffer
        return;
      }).on('end', function() {

        body = JSON.parse(body);
        body.readme = null; // remove this, it's too big

        // badge
        if (e === 'badge.svg') { // build another request
          var c = Object.keys(body.versions).length;
          var plu = c > 1 ? 's-' : '-';
          return http.get({
            host: 'img.shields.io',
            path: '/badge/version' + plu + c + '-red.svg' + s,
            agent: httpAgent,
            headers: {
              'User-Agent': VERSION
            }
          }, function(inp) {

            if (inp.statusCode !== 200) {
              return next(new Error(status[404]));
            }
            var badge = new Buffer(0);

            return inp.on('data', function(chunk) {

              badge += chunk; // buffer
              return;
            }).on('end', function() {

              var content = 'image/svg+xml; charset=utf-8';
              if (setHeader(res, 'Content-Type', content) === true) {
                res.send(badge);
                cache(badge, hash, content);
              }
              return;
            }).on('error', function(err) {

              next(new Error(status[404]));
              return debug('client', {
                pid: process.pid,
                status: 'response',
                error: err.message
              });
            });
          }).on('error', function(err) {

            next(new Error(status[404]));
            return debug('client', {
              pid: process.pid,
              status: 'response',
              error: err.message
            });
          });
        } // EOF badge

        var content = 'application/json; charset=utf-8';
        if (setHeader(res, 'Content-Type', content) === true) {
          res.send(body);
          cache(body, hash, content);
        }
        return;
      });
    }).on('error', function(err) {

      next(new Error(status[404]));
      return debug('client', {
        pid: process.pid,
        status: 'request',
        error: err.message
      });
    });

  });
  /**
   * index
   * 
   * @param {Object} req - client request
   * @param {Object} res - response to client
   */
  app.get('/', function(req, res) {

    return res.sendFile(index);
  });
  /**
   * catch all errors returned from page
   * 
   * @function
   * @param {Object} err - error raised
   * @param {Object} req - client request
   * @param {Object} res - response to client
   * @param {next} next - next callback
   */
  app.use(function(err, req, res, next) {

    var out = '';
    var error = err.message.toLowerCase(); // string
    var code = 500;
    debug('web', {
      pid: process.pid,
      status: 'catch',
      error: error
    });
    switch (error) {
      case 'not found':
        return next();
      default:
        if (my.env !== 'production') {
          out = error;
        }
        break;
    }
    res.status(code).json({
      error: out
    });
    return;
  });
  /**
   * catch error 404 or if nobody cannot parse the request
   * 
   * @function
   * @param {Object} req - client request
   * @param {Object} res - response to client
   */
  app.use(function(req, res) {

    var code = 404;
    res.status(code).json({
      error: status[code].toLowerCase()
    });
    return;
  });

  if (my.env != 'test') {
    if (my.env != 'production') {
      console.log(process.pid + ' | listening on: ' + my.host + ':' + my.port);
    }
    debug('web', {
      pid: process.pid,
      host: my.host,
      port: my.port
    });
    var server = http.createServer(app);
    if (my.timeout && my.timeout.milliseconds) {
      server.timeout = my.timeout.milliseconds;
    }
    server.listen(my.port, my.host);
  }
  return app;
}

/**
 * option setting
 * 
 * @exports supergiovane
 * @function supergiovane
 * @param {Object} opt - various options. Check README.md
 * @return {Object}
 */
module.exports = function supergiovane(opt) {

  var options = opt || Object.create(null);
  var my = {
    env: String(options.env || 'production'),
    host: String(options.host || '127.0.0.1'),
    port: Number(options.port) || 3000,
    referer: new RegExp(String(options.referer || 'http://127.0.0.1'), 'i'),
    dir: String(options.dir || __dirname + '/public/'),
    logger: options.logger === false ? false : options.logger || {
      filename: 'route.log'
    },
    timeout: options.timeout === false ? false : options.timeout || {},
    sitemap: options.sitemap === false ? false : options.sitemap || {},
    signature: options.signature === false ? false : options.signature || false,
    cache: options.cache === false ? false : Number(options.cache) || 6,
    flush: Number(options.flush) || 86400000,
    fork: Number(opt.fork) < 5 ? require('os').cpus().length : Number(opt.fork)
      || require('os').cpus().length,
    max: typeof opt.max === 'string' ? opt.max : Number(opt.max) || 0,
    debug: options.debug === false ? false : options.debug || 'debug.log',
    task: Boolean(options.task) ? options.task : false,
    mamma: Boolean(options.mamma) ? options.mamma : false
  };
  if (my.debug) {
    debug = logger({
      filename: my.debug,
      standalone: true,
      winston: {
        logger: '_spDebug',
        level: 'debug',
        json: false
      }
    });
  }

  /*
   * father
   */
  if (cluster.isMaster) {
    if (my.task) {
      require('task-manager')(my.task);
    }

    // no cluster
    if (my.env != 'production') {
      return bootstrap(my);
    }

    // cluster
    for (var i = 0; i < my.fork; i++) {
      cluster.fork();
    }
    /**
     * work if child die
     * 
     * @function
     * @param {Object} worker - worker child
     * @param {Integer} code - error code
     * @param {String} signal - error signal
     */
    cluster.on('exit', function(worker, code, signal) {

      debug('cluster', {
        pid: worker.process.pid,
        status: code || signal,
        suicide: worker.suicide,
        max: my.max
      });
      if (worker.suicide === true) { // task-manager kill or disconnect
        cluster.fork();
      } else if (isNaN(my.max) === true || my.max-- > 0) { // bug restart, not too much
        cluster.fork();
      }
      return;
    });
    return;
  }

  /*
   * child
   */
  process.on('uncaughtException', function(err) {

    debug('cluster', {
      pid: process.pid,
      status: 'uncaughtException',
      error: err.message,
      stack: err.stack
    });
    return setTimeout(function() { // wait logger write

      return process.exit(1);
    }, 250);
  });
  return bootstrap(my);
};
